### ISSUE TITLE
GPU MUCH slower than CPU

### ISSUE THREAD

**Issue by camj256** [tensorflow/tensorflow#5995]
GPU training is MUCH slower than CPU training. It's possible I'm doing something wrong. If I'm not I can gather more data on this. The data set is pretty small and it slows to a crawl. GPU usage is around 2-5%, It fills up the memory in the GPU pretty quickly to 90% but the PCIe Bandwidth Utilization is 1%. My CPU and Memory usage are otherwise minimal.

My setup: 32gb ram, 8 core 4.3 Ghz processor, (2) GTX 660's, 367.57 Nvidia Driver, Cuda Toolkit 7.5, cudnn 7.5, Python 2.7. Tensorflow matches 7.5.

I can take exact time measurements later if needed but I would guess GPU is about 10x slower if not more than the CPU training.

```
pprint.pprint(len(X))
pprint.pprint(len(Y))

net = tflearn.input_data(shape=[None, 7])
net = tflearn.fully_connected(net, 32)
net = tflearn.fully_connected(net, 32)
net = tflearn.fully_connected(net, 2, activation='softmax')
net = tflearn.regression(net)

model = tflearn.DNN(net)
model.fit(X, Y, n_epoch=100, batch_size=16, show_metric=True)

model.save("model.tfl")
```

```
camj256@camj256:~/PycharmProjects/DeepLearning$ python build.py 
I tensorflow/stream_executor/dso_loader.cc:108] successfully opened CUDA library libcublas.so locally
I tensorflow/stream_executor/dso_loader.cc:108] successfully opened CUDA library libcudnn.so locally
I tensorflow/stream_executor/dso_loader.cc:108] successfully opened CUDA library libcufft.so locally
I tensorflow/stream_executor/dso_loader.cc:108] successfully opened CUDA library libcuda.so.1 locally
I tensorflow/stream_executor/dso_loader.cc:108] successfully opened CUDA library libcurand.so locally
Exceptions: 0
3003
3003
I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:925] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
I tensorflow/core/common_runtime/gpu/gpu_init.cc:102] Found device 0 with properties: 
name: GeForce GTX 660
major: 3 minor: 0 memoryClockRate (GHz) 1.0975
pciBusID 0000:02:00.0
Total memory: 1.99GiB
Free memory: 1.27GiB
W tensorflow/stream_executor/cuda/cuda_driver.cc:572] creating context when one is currently active; existing: 0x2af5d00
I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:925] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
I tensorflow/core/common_runtime/gpu/gpu_init.cc:102] Found device 1 with properties: 
name: GeForce GTX 660
major: 3 minor: 0 memoryClockRate (GHz) 1.0975
pciBusID 0000:01:00.0
Total memory: 1.99GiB
Free memory: 1.24GiB
I tensorflow/core/common_runtime/gpu/gpu_init.cc:59] cannot enable peer access from device ordinal 0 to device ordinal 1
I tensorflow/core/common_runtime/gpu/gpu_init.cc:59] cannot enable peer access from device ordinal 1 to device ordinal 0
I tensorflow/core/common_runtime/gpu/gpu_init.cc:126] DMA: 0 1 
I tensorflow/core/common_runtime/gpu/gpu_init.cc:136] 0:   Y N 
I tensorflow/core/common_runtime/gpu/gpu_init.cc:136] 1:   N Y 
I tensorflow/core/common_runtime/gpu/gpu_device.cc:838] Creating TensorFlow device (/gpu:0) -> (device: 0, name: GeForce GTX 660, pci bus id: 0000:02:00.0)
I tensorflow/core/common_runtime/gpu/gpu_device.cc:838] Creating TensorFlow device (/gpu:1) -> (device: 1, name: GeForce GTX 660, pci bus id: 0000:01:00.0)
I tensorflow/core/common_runtime/gpu/gpu_device.cc:838] Creating TensorFlow device (/gpu:0) -> (device: 0, name: GeForce GTX 660, pci bus id: 0000:02:00.0)
I tensorflow/core/common_runtime/gpu/gpu_device.cc:838] Creating TensorFlow device (/gpu:1) -> (device: 1, name: GeForce GTX 660, pci bus id: 0000:01:00.0)
---------------------------------
Run id: 17E6I9
Log directory: /tmp/tflearn_logs/
---------------------------------
Training samples: 3003
Validation samples: 0
```

**Comment by poxvoculi** [comment]
Have you tried using timeline to find out where the time is going?

http://stackoverflow.com/questions/34293714/tensorflow-can-i-measure-the-execution-time-of-individual-operations/37774470#37774470

**Comment by camj256** [comment]
This can be closed pretty soon if nothing jumps out at anyone. I screwed up my linux install late last night messing with drivers. Once I have time to fix my linux install and dig into this I can re-open.

I'm wondering if the batch size was too small and it was bottle necking somewhere?

**Comment by Mazecreator** [comment]
@camj256,

I have similar issues I still have not got to the bottom of with an RL model.  It seems like TensorFlow maybe optimized for larger datasets, but my understanding is other deep learning tools might handle the smaller datasets more efficiently.  I would rather stay with TensorFlow if I can so it would be nice if some optimization can be done on smaller datasets to get close to the performance of the other tools.  There is an open item #2444 which also has documented the same issue.  I have a StackOverflow posting as well to see if it is just a matter of optimization.
http://stackoverflow.com/questions/38688777/tensorflow-graph-optimization-gpu-vs-cpu-performance

Hopefully we can move TensorFlow in the director of some optimization for the smaller dataset sizes.

**Comment by camj256** [comment]
My actual data set is 200k. Same result there too.

**Comment by poxvoculi** [comment]
The basic question here is whether your poor performance is due to a slow GPU kernel, to inefficiencies in moving data back and forth between CPU and GPU, or something else.   If you could use timeline and other profiling tools to track this down on your side, it would be helpful.   When data sizes are tiny, GPUs provide no advantage over CPUs.  However, if another DNN framework can run this same model on GPU much faster, that's a performance problem we'd like to address.

**Comment by camj256** [comment]
I'm gonna close this for now then. I won't be able to look into that for another week or so. Too busy with work.

**Comment by camj256** [comment]
I got my linux install fixed and am looking into this right now. Hopefully someone can answer a few rookie questions that are getting in the way of my understanding of this.

I've noticed that I can just up the batch size to 100k and process a lot quicker than CPU but the question is does the batch size have any affect on how the DNN learns? Also, I've noticed if I change the size of the NN to be much larger like 512 instead of 32 it seems to just get stuck at the same acc forever BUT when doing so the GPU usage goes up to like 50% on one of the cards. 

I'm going to run the timeline thing later tonight but for now the above questions would help me understand what I'm doing a lot better. (Sorry, I've tried googling and the docs on the site seem to assume I already know what everything does.)

**Comment by poxvoculi** [comment]
There's no simple answer to these questions.  But there should be many papers and blog posts about the general topics of SGD training and hyperparameter tuning.

If you're really a beginner, you should try playing with http://playground.tensorflow.org/

**Comment by camj256** [comment]
Well, I went and bought a gtx 1080 tonight. It was time to upgrade. Turns out this most likely is a bug as my 1080 hauls absolute ass with the exact same code and setup (even same drivers) 

As for what it was I will never know. Probably just CUDA the 660's being an older version of CUDA?
